File: backend/app/schemas/validation.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator

from app.domain import AssignmentRole

ValidationStudyStatus = Literal["draft", "active", "frozen", "completed", "archived"]
ValidationPartition = Literal["development", "internal_test", "external_test", "shadow"]
ValidationCaseStatus = Literal["queued", "under_review", "ready", "compared", "excluded"]


class ValidationStudyCreateRequest(BaseModel):
    name: str = Field(min_length=3, max_length=300)
    description: str | None = Field(default=None, max_length=5000)
    scheme_code: str = Field(default="MOC-ST", min_length=2, max_length=50)
    protocol_version: str = Field(
        default="expert-grounded-validation-v1", min_length=3, max_length=50
    )
    annotation_rulebook_version: str = Field(
        default="expert-annotation-rulebook-v1", min_length=3, max_length=50
    )
    shadow_mode: bool = True
    minimum_reviews_per_case: int = Field(default=2, ge=2, le=10)
    recommendation_policy: dict[str, Any] = Field(default_factory=dict)

    @field_validator("scheme_code")
    @classmethod
    def normalize_scheme_code(cls, value: str) -> str:
        return value.strip().upper()


class ValidationStudyStatusRequest(BaseModel):
    status: ValidationStudyStatus


class ValidationCaseCreateRequest(BaseModel):
    proposal_id: str = Field(min_length=1)
    proposal_version_number: int | None = Field(default=None, ge=1)
    partition: ValidationPartition = "shadow"


class ValidationCaseExcludeRequest(BaseModel):
    reason: str = Field(min_length=10, max_length=2000)

    @field_validator("reason")
    @classmethod
    def strip_reason(cls, value: str) -> str:
        return value.strip()


class ValidationAssignmentCreateRequest(BaseModel):
    reviewer_email: str = Field(min_length=3, max_length=320)
    role: AssignmentRole = AssignmentRole.TECHNICAL

    @field_validator("reviewer_email")
    @classmethod
    def normalize_email(cls, value: str) -> str:
        cleaned = value.strip().lower()
        if "@" not in cleaned or cleaned.startswith("@") or cleaned.endswith("@"):
            raise ValueError("A valid reviewer email is required")
        return cleaned


class ValidationStudyResponse(BaseModel):
    id: str
    name: str
    description: str | None
    scheme_id: str
    scheme_code: str
    rubric_version_id: str
    rubric_version: str
    model_version_id: str
    model_name: str
    model_version: str
    model_artifact_hash: str
    rubric_definition_hash: str
    protocol_version: str
    annotation_rulebook_version: str
    status: str
    shadow_mode: bool
    minimum_reviews_per_case: int
    recommendation_policy: dict[str, Any]
    created_by: str
    created_at: datetime
    activated_at: datetime | None
    frozen_at: datetime | None
    completed_at: datetime | None
    case_count: int = 0
    compared_case_count: int = 0


class ValidationStudyListResponse(BaseModel):
    studies: list[ValidationStudyResponse]


class ValidationCaseResponse(BaseModel):
    id: str
    study_id: str
    proposal_id: str
    proposal_version_id: str
    proposal_version_number: int
    proposal_title: str
    model_run_id: str
    partition: str
    status: str
    exclusion_reason: str | None
    included_by: str
    included_at: datetime
    comparison_ready_at: datetime | None
    assigned_reviewers: int = 0
    completed_reviews: int = 0
    minimum_reviews_required: int = 2
    model_output_blinded: bool = True


class ValidationCaseListResponse(BaseModel):
    study_id: str
    cases: list[ValidationCaseResponse]


class ValidationCriterionFormItem(BaseModel):
    criterion_id: str
    criterion_key: str | None
    category: str
    criterion: str
    maximum: float
    description: str | None
    order: int


class ValidationReviewFormResponse(BaseModel):
    assignment_id: str
    proposal_id: str
    proposal_version_id: str
    proposal_version_number: int
    proposal_title: str
    reviewer_role: str
    validation_case_id: str | None
    study_name: str | None
    protocol_version: str | None
    annotation_rulebook_version: str | None
    shadow_mode: bool
    model_output_hidden: bool
    rubric_version: str
    total_marks: int
    criteria: list[ValidationCriterionFormItem]


class ValidationMetricResponse(BaseModel):
    name: str
    value: float | None
    sample_size: int
    details: dict[str, Any] = Field(default_factory=dict)


class ValidationReadinessResponse(BaseModel):
    scientifically_validated: bool = False
    status: str
    warnings: list[str]
    total_cases: int
    compared_cases: int
    completed_reviews: int
    minimum_reviews_per_case: int
    partitions: dict[str, int]


class ValidationStudySummaryResponse(BaseModel):
    study: ValidationStudyResponse
    readiness: ValidationReadinessResponse
    snapshot_group_id: str | None
    metrics: list[ValidationMetricResponse]
    computed_at: datetime | None


class ValidationComputeResponse(BaseModel):
    study_id: str
    snapshot_group_id: str
    compared_cases: int
    metrics_written: int
    warnings: list[str]
    message: str


class ValidationAssignmentResponse(BaseModel):
    assignment_id: str
    validation_case_id: str
    reviewer_id: str
    status: str
    blind: bool
    message: str

File: src/api/validation.ts
import { apiDownload, apiFetch } from "./client";

export interface ValidationStudyResponse {
  id: string;
  name: string;
  description?: string | null;
  scheme_id: string;
  scheme_code: string;
  rubric_version_id: string;
  rubric_version: string;
  model_version_id: string;
  model_name: string;
  model_version: string;
  model_artifact_hash: string;
  rubric_definition_hash: string;
  protocol_version: string;
  annotation_rulebook_version: string;
  status: string;
  shadow_mode: boolean;
  minimum_reviews_per_case: number;
  recommendation_policy: Record<string, unknown>;
  created_by: string;
  created_at: string;
  activated_at?: string | null;
  frozen_at?: string | null;
  completed_at?: string | null;
  case_count: number;
  compared_case_count: number;
}

export interface ValidationStudyListResponse {
  studies: ValidationStudyResponse[];
}

export interface ValidationCaseResponse {
  id: string;
  study_id: string;
  proposal_id: string;
  proposal_version_id: string;
  proposal_version_number: number;
  proposal_title: string;
  model_run_id: string;
  partition: string;
  status: string;
  exclusion_reason?: string | null;
  included_by: string;
  included_at: string;
  comparison_ready_at?: string | null;
  assigned_reviewers: number;
  completed_reviews: number;
  minimum_reviews_required: number;
  model_output_blinded: boolean;
}

export interface ValidationCaseListResponse {
  study_id: string;
  cases: ValidationCaseResponse[];
}

export interface ValidationMetricResponse {
  name: string;
  value?: number | null;
  sample_size: number;
  details: Record<string, unknown>;
}

export interface ValidationReadinessResponse {
  scientifically_validated: boolean;
  status: string;
  warnings: string[];
  total_cases: number;
  compared_cases: number;
  completed_reviews: number;
  minimum_reviews_per_case: number;
  partitions: Record<string, number>;
}

export interface ValidationStudySummaryResponse {
  study: ValidationStudyResponse;
  readiness: ValidationReadinessResponse;
  snapshot_group_id?: string | null;
  metrics: ValidationMetricResponse[];
  computed_at?: string | null;
}

export interface ValidationComputeResponse {
  study_id: string;
  snapshot_group_id: string;
  compared_cases: number;
  metrics_written: number;
  warnings: string[];
  message: string;
}

export interface ValidationCriterionFormItem {
  criterion_id: string;
  criterion_key?: string | null;
  category: string;
  criterion: string;
  maximum: number;
  description?: string | null;
  order: number;
}

export interface ValidationReviewFormResponse {
  assignment_id: string;
  proposal_id: string;
  proposal_version_id: string;
  proposal_version_number: number;
  proposal_title: string;
  reviewer_role: string;
  validation_case_id?: string | null;
  study_name?: string | null;
  protocol_version?: string | null;
  annotation_rulebook_version?: string | null;
  shadow_mode: boolean;
  model_output_hidden: boolean;
  rubric_version: string;
  total_marks: number;
  criteria: ValidationCriterionFormItem[];
}

export async function listValidationStudies(): Promise<ValidationStudyListResponse> {
  return apiFetch<ValidationStudyListResponse>("/validation/studies");
}

export async function createValidationStudy(body: {
  name: string;
  description?: string | null;
  scheme_code: string;
  protocol_version: string;
  annotation_rulebook_version: string;
  shadow_mode: boolean;
  minimum_reviews_per_case: number;
  recommendation_policy?: Record<string, unknown>;
}): Promise<ValidationStudyResponse> {
  return apiFetch<ValidationStudyResponse>("/validation/studies", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export async function getValidationStudy(
  studyId: string,
): Promise<ValidationStudySummaryResponse> {
  return apiFetch<ValidationStudySummaryResponse>(`/validation/studies/${studyId}`);
}

export async function updateValidationStudyStatus(
  studyId: string,
  status: string,
): Promise<ValidationStudyResponse> {
  return apiFetch<ValidationStudyResponse>(
    `/validation/studies/${studyId}/status`,
    { method: "PATCH", body: JSON.stringify({ status }) },
  );
}

export async function listValidationCases(
  studyId: string,
): Promise<ValidationCaseListResponse> {
  return apiFetch<ValidationCaseListResponse>(
    `/validation/studies/${studyId}/cases`,
  );
}

export async function addValidationCase(
  studyId: string,
  body: {
    proposal_id: string;
    proposal_version_number?: number | null;
    partition: string;
  },
): Promise<ValidationCaseResponse> {
  return apiFetch<ValidationCaseResponse>(`/validation/studies/${studyId}/cases`, {
    method: "POST",
    body: JSON.stringify(body),
  });
}

export async function excludeValidationCase(
  caseId: string,
  reason: string,
): Promise<ValidationCaseResponse> {
  return apiFetch<ValidationCaseResponse>(
    `/validation/cases/${caseId}/exclude`,
    { method: "PATCH", body: JSON.stringify({ reason }) },
  );
}

export async function assignShadowReviewer(
  caseId: string,
  reviewerEmail: string,
  role: "technical" | "financial",
): Promise<{ assignment_id: string; message: string }> {
  return apiFetch<{ assignment_id: string; message: string }>(
    `/validation/cases/${caseId}/assignments`,
    {
      method: "POST",
      body: JSON.stringify({ reviewer_email: reviewerEmail, role }),
    },
  );
}

export async function getValidationReviewForm(
  assignmentId: string,
): Promise<ValidationReviewFormResponse> {
  return apiFetch<ValidationReviewFormResponse>(
    `/validation/assignments/${assignmentId}/form`,
  );
}

export async function computeValidationMetrics(
  studyId: string,
): Promise<ValidationComputeResponse> {
  return apiFetch<ValidationComputeResponse>(
    `/validation/studies/${studyId}/compute`,
    { method: "POST" },
  );
}

export async function downloadValidationDataset(
  studyId: string,
  includeEvidence = false,
): Promise<Blob> {
  return apiDownload(
    `/validation/studies/${studyId}/export?include_evidence=${includeEvidence}`,
  );
}
